tests/kernel/mem_protect/mpu/src/inherit.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "test_mpu.h"
+
+/* function prototypes */
+static inline void dummy_start(struct k_timer *timer){ ARG_UNUSED(timer);}
+static inline void dummy_end(struct k_timer *timer){ ARG_UNUSED(timer);}
+
+/* Kernel objects */
+K_THREAD_STACK_DEFINE(test_1_stack, INHERIT_STACK_SIZE);
+K_SEM_DEFINE(inherit_sem, SEMAPHORE_INIT_COUNT, SEMAPHORE_MAX_COUNT);
+K_MUTEX_DEFINE(inherit_mutex);
+K_TIMER_DEFINE(inherit_timer, dummy_start, dummy_end);
+K_MSGQ_DEFINE(inherit_msgq, MSG_Q_SIZE, MSG_Q_MAX_NUM_MSGS, MSG_Q_ALIGN);
+__kernel struct k_thread test_1_tid;
+
+u8_t MEM_DOMAIN_ALIGNMENT inherit_buf[MEM_REGION_ALLOC]; /* for mem domain */
+
+K_MEM_PARTITION_DEFINE(inherit_memory_partition,
+		       inherit_buf,
+		       sizeof(inherit_buf),
+		       K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *inherit_memory_partition_array[] = {
+	&inherit_memory_partition
+};
+
+struct k_mem_domain inherit_mem_domain;
+
+/* generic function to do check the access permissions. */
+void access_test(void)
+{
+	u32_t msg_q_data = 0xA5A5;
+	/* check for all accesses  */
+	k_sem_give(&inherit_sem);
+	k_mutex_lock(&inherit_mutex,K_FOREVER);
+	(void) k_timer_status_get(&inherit_timer);
+	k_msgq_put(&inherit_msgq, (void *)&msg_q_data, K_NO_WAIT);
+	k_mutex_unlock(&inherit_mutex);
+	inherit_buf[10] = 0xA5;
+}
+
+void test_thread_1_for_user(void *p1, void *p2, void *p3)
+{
+	access_test();
+	ztest_test_pass();
+}
+
+void test_thread_1_for_SU(void *p1, void *p2, void *p3)
+{
+	valid_fault = false;
+	access_test();
+
+	/* Check if user mode inherit is working if control is passed from SU */
+	k_thread_user_mode_enter(test_thread_1_for_user, NULL, NULL, NULL);
+}
+
+void inherit_test_1(void *p1, void *p2, void *p3)
+{
+
+	k_mem_domain_init(&inherit_mem_domain,
+			  1,
+			  inherit_memory_partition_array);
+
+	k_mem_domain_add_thread(&inherit_mem_domain, k_current_get());
+
+	k_thread_access_grant(k_current_get(),
+			      &inherit_sem,
+			      &inherit_mutex,
+			      &inherit_timer,
+			      &inherit_msgq, &test_1_stack, NULL);
+
+	k_thread_create(&test_1_tid,
+			test_1_stack,
+			INHERIT_STACK_SIZE,
+			test_thread_1_for_SU,
+			NULL, NULL, NULL,
+			0, K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_sleep(1000);
```

Comment:
The only parts of this test that don't seem to already be covered by existing tests (e.g. tests/kernel/{msgq, sem, mutex} all rely on K_INHERIT_PERMS working and would fail otherwise, as does ztest itself) is the memory domain test and the use of k_thread_user_mode_enter(), which could be tested independently of each other just by adding two tests to userspace.  We could also do the setup of the memory domain in the test_main() function and then just directly use a ztest_user_unit_test() instead of creating multiple layers of threads, needing to yield, etc. ztest is itself creating a thread with K_INHERIT_PERMS, so it is already exercising inheritance of anything set up by test_main().